compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -1608,131 +1608,124 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                        param_env,
+                        obligations,
+                        candidates,
+                    )
+                })?;
+
+                self.check_assoc_ty(assoc_item, name, def_scope, block, span);
+
+                // FIXME(fmease): Currently creating throwaway `parent_args` to please
+                // `create_args_for_associated_item`. Modify the latter instead (or sth. similar) to
+                // not require the parent args logic.
+                let parent_args = ty::GenericArgs::identity_for_item(tcx, impl_);
+                let args =
+                    self.create_args_for_associated_item(span, assoc_item, segment, parent_args);
+                let args = tcx.mk_args_from_iter(
+                    std::iter::once(ty::GenericArg::from(self_ty))
+                        .chain(args.into_iter().skip(parent_args.len())),
+                );
 
-                fn fold_const(
-                    &mut self,
-                    ct: ty::Const<'tcx>,
-                ) -> <TyCtxt<'tcx> as rustc_type_ir::Interner>::Const {
-                    assert!(!ct.ty().has_escaping_bound_vars());
-
-                    match ct.kind() {
-                        ty::ConstKind::Bound(_, bv) => ty::Const::new_placeholder(
-                            self.tcx,
-                            ty::PlaceholderConst { universe: self.universe, bound: bv },
-                            ct.ty(),
-                        ),
-                        _ => ct.super_fold_with(self),
-                    }
-                }
-            }
+                let ty = Ty::new_alias(tcx, ty::Inherent, ty::AliasTy::new(tcx, assoc_item, args));
 
-            let InferOk { value: self_ty, obligations } =
-                infcx.at(&cause, param_env).normalize(self_ty);
+                Ok(Some((ty, assoc_item)))
+            },
+        )
+    }
 
-            candidates
-                .iter()
-                .copied()
-                .filter(|&(impl_, _)| {
-                    infcx.probe(|_| {
-                        let ocx = ObligationCtxt::new(infcx);
-                        ocx.register_obligations(obligations.clone());
-
-                        let impl_args = infcx.fresh_args_for_item(span, impl_);
-                        let impl_ty = tcx.type_of(impl_).instantiate(tcx, impl_args);
-                        let impl_ty = ocx.normalize(&cause, param_env, impl_ty);
-
-                        // Check that the self types can be related.
-                        if ocx.eq(&ObligationCause::dummy(), param_env, impl_ty, self_ty).is_err() {
-                            return false;
-                        }
+    fn select_inherent_assoc_type_candidates(
+        &self,
+        infcx: &InferCtxt<'tcx>,
+        name: Ident,
+        span: Span,
+        self_ty: Ty<'tcx>,
+        cause: ObligationCause<'tcx>,
+        param_env: ParamEnv<'tcx>,
+        obligations: Vec<Obligation<'tcx, Predicate<'tcx>>>,
```

Comment:
Remove these obligations, and normalize inside of the probe inside of this function.

```diff
  let ocx = ObligationCtxt::new(infcx);
- ocx.register_obligations(obligations.clone());
+ let self_ty = ocx.normalize(&ObligationCause::dummy(), param_env, self_ty);
```